### PR TITLE
Add map drawing method like original style

### PR DESCRIPTION
**Pull request by kunyihua**
1. Add MapPosition.Center
2. Add noShift option in Compose
3. If use MapPosition.Center, The map will be draw like original style

### CODE REVIEW

File: Helpers/Compositor.cs [hunk 1]
```diff
@@ -29,8 +29,10 @@ namespace D2RAssist.Helpers
     public class Compositor
     {
         private readonly AreaData _areaData;
+        public AreaData AreaData { get { return _areaData; } }
         private readonly Bitmap _background;
         private readonly Point _cropOffset;
+        public Point CropOffset { get { return _cropOffset; } }
```
Comment (AudriusButkevicius) [inline Helpers/Compositor.cs:35]: Remove, make:
```
private readonly Point _cropOffset;
```
into
```
public readonly Point CropOffset;
```

File: Overlay.cs [hunk 2]
```diff
@@ -155,19 +156,66 @@ private void MapOverlay_Paint(object sender, PaintEventArgs e)
...
+                var p1 = Transform(new Vector2(0, 0));
+                var p2 = Transform(new Vector2(gameMap.Width, 0));
+                var p3 = Transform(new Vector2(gameMap.Width, gameMap.Height));
+                var p4 = Transform(new Vector2(0, gameMap.Height));
+
+                System.Drawing.PointF[] destinationPoints = {
+                        new System.Drawing.PointF(p1.X, p1.Y),
+                        new System.Drawing.PointF(p2.X, p2.Y),
+                        new System.Drawing.PointF(p4.X, p4.Y)
+                    };
```
Comment (misterokaygo) [inline Overlay.cs:189]: Is there extra indentation here?

**Review by misterokaygo — CHANGES_REQUESTED** [review]

File: Helpers/Compositor.cs (hunk 1 above)
Comment (kunyihua) [reply Helpers/Compositor.cs:35]: _cropOffset will be used in many places in this file...So I think it’s best to add a public variable and return the value directly

Comment (kunyihua) [reply Helpers/Compositor.cs]: Thank you, I modified it like this
`public readonly Point _cropOffset;`

File: Helpers/Compositor.cs [hunk 3]
```diff
@@ -44,7 +46,7 @@ public Compositor(AreaData areaData, IReadOnlyList<PointOfInterest> pointOfInter
             (_background, _cropOffset) = DrawBackground(areaData, pointOfInterest);
         }
 
-        public Bitmap Compose(GameData gameData)
+        public Bitmap Compose(GameData gameData, bool noShift = false)
```
[the remark this answers is not in the document]
Comment (kunyihua) [reply Helpers/Compositor.cs]: I fixed it

File: Helpers/Compositor.cs [hunk 4]
```diff
@@ -29,8 +29,10 @@ namespace D2RAssist.Helpers
     public class Compositor
     {
         private readonly AreaData _areaData;
+        public AreaData AreaData { get { return _areaData; } }
```
[the remark this answers is not in the document]
Comment (kunyihua) [reply Helpers/Compositor.cs]: Thank you, I modified it like this
`public readonly AreaData _areaData;`

File: Overlay.cs [hunk 5]
```diff
@@ -155,19 +156,66 @@ private void MapOverlay_Paint(object sender, PaintEventArgs e)
...
+
+            if (Settings.Map.Position == MapPosition.Center)
+            {
+                float w = this.Width;
+                float h = this.Height;
+
+                int s_p_offset_x = _currentGameData.PlayerPosition.X - _compositor.AreaData.Origin.X - _compositor.CropOffset.X;
+                int s_p_offset_y = _currentGameData.PlayerPosition.Y - _compositor.AreaData.Origin.Y - _compositor.CropOffset.Y;
+
+                Vector2 _screenCenterCache = new Vector2(w / 2, h / 2);
```
[the remark this answers is not in the document]
Comment (kunyihua) [reply Overlay.cs]: I fixed it

File: Overlay.cs (hunk 2 above)
Comment (kunyihua) [reply Overlay.cs:215]: What do you mean, I don't understand?

File: Overlay.cs [hunk 6]
```diff
@@ -155,19 +157,96 @@ private void MapOverlay_Paint(object sender, PaintEventArgs e)
...
 
-            Bitmap gameMap = _compositor.Compose(_currentGameData);
-            var anchor = new Point(0, 0);
-            switch (Settings.Map.Position)
+            Bitmap gameMap = _compositor.Compose(_currentGameData, true);
+
+            float w = 0;
+            float h = 0;
+            float scale = 0.0F;
+            Vector2 _screenCenterCache = new Vector2();
```
Comment (AudriusButkevicius) [inline Overlay.cs:165]: Please rename this to `center` without `_`.
`_` is only used for class member variables.

File: Overlay.cs [hunk 7]
```diff
@@ -155,19 +157,96 @@ private void MapOverlay_Paint(object sender, PaintEventArgs e)
...
+            //        case MapPosition.TopRight:
+            //            anchor = new Point(_screen.WorkingArea.Width - gameMap.Width, 0);
+            //            break;
+            //        case MapPosition.TopLeft:
+            //            anchor = new Point(0, 0);
+            //            break;
+            //    }
+
+            //    e.Graphics.DrawImage(gameMap, anchor);
+            //}
```
Comment (AudriusButkevicius) [inline Overlay.cs:238]: Please remove the dead code.

File: Overlay.cs [hunk 8]
```diff
@@ -155,19 +157,96 @@ private void MapOverlay_Paint(object sender, PaintEventArgs e)
 
             UpdateLocation();
 
-            Bitmap gameMap = _compositor.Compose(_currentGameData);
-            var anchor = new Point(0, 0);
-            switch (Settings.Map.Position)
+            Bitmap gameMap = _compositor.Compose(_currentGameData, true);
```
Comment (AudriusButkevicius) [inline Overlay.cs:160]: This is always called with `true`, which seems wrong?

I assumed that this feature was supposed to be behind a feature flag?

**Review by misterokaygo — APPROVED** [review]

File: Helpers/Compositor.cs [hunk 9]
```diff
@@ -30,7 +30,7 @@ public class Compositor
     {
         private readonly AreaData _areaData;
         private readonly Bitmap _background;
-        private readonly Point _cropOffset;
+        public readonly Point _cropOffset;
```
Comment (AudriusButkevicius) [inline Helpers/Compositor.cs:33]: Change this to `CropSize` as this is now public.

File: Overlay.cs [hunk 10]
```diff
@@ -155,19 +157,97 @@ private void MapOverlay_Paint(object sender, PaintEventArgs e)
...
+                    center +
+                    DeltaInWorldToMinimapDelta(
+                        p - playerPos,
+                        (float)Math.Sqrt(w * w + h * h),
+                        scale,
+                        0);
+
+                var p1 = Transform(new Vector2(0, 0));
+                var p2 = Transform(new Vector2(gameMap.Width, 0));
+                var p3 = Transform(new Vector2(gameMap.Width, gameMap.Height));
```
Comment (AudriusButkevicius) [inline Overlay.cs:211]: Never used?

File: Overlay.cs [hunk 11]
```diff
@@ -155,19 +157,97 @@ private void MapOverlay_Paint(object sender, PaintEventArgs e)
...
+
+                e.Graphics.DrawImage(gameMap, destinationPoints);
+            }
+            else
+            {
+                var anchor = new Point(0, 0);
+                switch (Settings.Map.Position)
+                {
+                    case MapPosition.Center:
+                        anchor = new Point(_screen.WorkingArea.Width / 2, _screen.WorkingArea.Height / 2);
```
Comment (AudriusButkevicius) [inline Overlay.cs:228]: We should probably still center around player in this mode?

**Pull request merged**